packages/reframed/src/reframed.ts

```diff
@@ -206,50 +190,42 @@ async function reframeFromTarget(
 	return promise;
 }
 
-function attachNavigationListeners(iframe: HTMLIFrameElement) {
-	const controller = new AbortController();
-	const signal = controller.signal;
-
-	// When a navigation event occurs on the main window, either programmatically through the History API or by the back/forward button,
-	// we need to reflect those changes onto the iframe's history via replaceState().
-	// replaceState() is used in order to avoid adding duplicate entries to the shared history stack.
-	// Finally, dispatch a PopStateEvent so that fragments that are listening to popstate event re-trigger render updates
-	const handleNavigate = () => {
-		iframe.contentWindow?.history.replaceState(
-			window.history.state,
-			"",
-			window.location.href
-		);
-		iframe.contentWindow?.dispatchEvent(new PopStateEvent("popstate"));
-	};
-
-	// reframed:navigate event is triggered by the patched main window.history methods
-	window.addEventListener("reframed:navigate", handleNavigate, { signal });
-
-	// Forward the popstate event triggered on the main window to every registered iframe window
-	window.addEventListener("popstate", handleNavigate, { signal });
-
-	// cleanup event listeners attached to the window when the iframe gets destroyed
-	iframe.contentWindow?.addEventListener("unload", () => controller.abort());
-}
-
 /**
  * Apply monkey-patches to the source iframe so that we trick code running in it to behave as if it
  * was running in the main frame.
  */
-function monkeyPatchIFrameDocument(
-	iframeDocument: Document,
+function monkeyPatchIFrameEnvironment(
+	iframe: HTMLIFrameElement,
 	shadowRoot: ReframedShadowRoot
 ) {
+	assert(
+		iframe.contentWindow !== null && iframe.contentDocument !== null,
+		"attempted to patch iframe before it was ready"
+	);
+
+	const iframeWindow = iframe.contentWindow as Window & typeof globalThis;
+	const iframeDocument = iframe.contentDocument;
+
+	// When an iframe element is added to the page, it always triggers a load event
+	// even if the src is empty. In this case, don't monkey patch the iframe document
+	// because we will end up doing it twice.
+	if (
```

Comment:
Not related to the PR, but I think this check will be unnecessary once we cleanup iframe initialization. Right now we set an initial empty src and then set equal to the reframedSrc after we fetch the reframed HTML. This results in 2 load events triggering, which I'm not sure we want to happen.

Instead, we should only set the src once. When we attempt to reframe from fetch, we should check the ready state of the iframe and if it's NOT ready, attach a handler to load event that will then stream HTML to main document through writeable-dom. If the iframe is ready, stream the HTML once the fetch resolves